**SUPERVISED LEARNING**

Mata Kuliah Analisis Data Eksploratif B083

Kelompok Fomo:
1. Karina Auralia (23083010072)
2. Marthalia Kusumarima (23083010100)

In [ ]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import matplotlib.pyplot as plt
import scipy as stats
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import accuracy_score, confusion_matrix, ConfusionMatrixDisplay, classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
df = pd.read_csv('weight_change_dataset.csv')
df

Participant ID  Age Gender  Current Weight (lbs)  BMR (Calories)  \
0                1   56      M                 228.4          3102.3   
1                2   46      F                 165.4          2275.5   
2                3   32      F                 142.8          2119.4   
3                4   25      F                 145.5          2181.3   
4                5   38      M                 155.5          2463.8   
..             ...  ...    ...                   ...             ...   
95              96   59      M                 181.9          2622.8   
96              97   56      F                 116.5          1736.5   
97              98   58      F                 162.7          2188.6   
98              99   45      M                 179.8          2671.9   
99             100   24      M                 185.2          2830.7   

    Daily Calories Consumed  Daily Caloric Surplus/Deficit  \
0                    3916.0                          813.7   
1                    3823.0                         1547.5   
2                    2785.4                          666.0   
3                    2587.3                          406.0   
4                    3312.8                          849.0   
..                      ...                            ...   
95                   3656.2                         1033.4   
96                   2756.5                         1020.0   
97                   3344.3                         1155.8   
98                   4000.0                         1328.1   
99                   4000.0                         1169.3   

    Weight Change (lbs)  Duration (weeks) Physical Activity Level  \
0               0.20000                 1               Sedentary   
1               2.40000                 6             Very Active   
2               1.40000                 7               Sedentary   
3               0.80000                 8               Sedentary   
4               2.00000                10          Lightly Active   
..                  ...               ...                     ...   
95              0.60000                 2          Lightly Active   
96             -8.92309                12       Moderately Active   
97              3.60000                12          Lightly Active   
98              2.00000                 5       Moderately Active   
99              2.70000                 9             Very Active   

   Sleep Quality  Stress Level  Final Weight (lbs)  
0      Excellent             6               228.6  
1      Excellent             6               167.8  
2           Good             3               144.2  
3           Fair             2               146.3  
4           Good             1               157.5  
..           ...           ...                 ...  
95     Excellent             3               182.5  
96          Good             9               107.6  
97          Fair             7               166.3  
98          Fair             5               181.8  
99          Fair             7               187.9  

[100 rows x 13 columns]

# Data Understanding

In [ ]:
# Memeriksa ukuran data
df.shape

(100, 13)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Participant ID                 100 non-null    int64  
 1   Age                            100 non-null    int64  
 2   Gender                         100 non-null    object 
 3   Current Weight (lbs)           100 non-null    float64
 4   BMR (Calories)                 100 non-null    float64
 5   Daily Calories Consumed        100 non-null    float64
 6   Daily Caloric Surplus/Deficit  100 non-null    float64
 7   Weight Change (lbs)            100 non-null    float64
 8   Duration (weeks)               100 non-null    int64  
 9   Physical Activity Level        100 non-null    object 
 10  Sleep Quality                  100 non-null    object 
 11  Stress Level                   100 non-null    int64  
 12  Final Weight (lbs)             100 non-null    floa

In [ ]:
# Memeriksa nama-nama kolom dalam dataset
df.columns

Index(['Participant ID', 'Age', 'Gender', 'Current Weight (lbs)',
       'BMR (Calories)', 'Daily Calories Consumed',
       'Daily Caloric Surplus/Deficit', 'Weight Change (lbs)',
       'Duration (weeks)', 'Physical Activity Level', 'Sleep Quality',
       'Stress Level', 'Final Weight (lbs)'],
      dtype='object')

In [ ]:
# Mmemeriksi apakah terdapat missing value dalam dataset
miss = df.isnull().sum().sort_values(ascending=False)
miss

Participant ID                   0
Age                              0
Gender                           0
Current Weight (lbs)             0
BMR (Calories)                   0
Daily Calories Consumed          0
Daily Caloric Surplus/Deficit    0
Weight Change (lbs)              0
Duration (weeks)                 0
Physical Activity Level          0
Sleep Quality                    0
Stress Level                     0
Final Weight (lbs)               0
dtype: int64

In [ ]:
df.describe()

Participant ID         Age  Current Weight (lbs)  BMR (Calories)  \
count      100.000000  100.000000            100.000000      100.000000   
mean        50.500000   37.910000            171.533000     2518.206000   
std         29.011492   12.219454             30.333856      364.431221   
min          1.000000   18.000000            100.000000     1566.500000   
25%         25.750000   26.750000            153.650000     2255.050000   
50%         50.500000   38.000000            172.150000     2519.500000   
75%         75.250000   46.250000            192.475000     2805.975000   
max        100.000000   59.000000            238.200000     3390.800000   

       Daily Calories Consumed  Daily Caloric Surplus/Deficit  \
count               100.000000                     100.000000   
mean               3518.292000                    1000.091000   
std                 513.313097                     371.560827   
min                2030.900000                      82.500000   
25%                3233.300000                     766.950000   
50%                3636.050000                    1013.100000   
75%                4000.000000                    1253.325000   
max                4000.000000                    1922.500000   

       Weight Change (lbs)  Duration (weeks)  Stress Level  Final Weight (lbs)  
count           100.000000        100.000000    100.000000          100.000000  
mean             -2.779817          6.920000      4.810000          168.750000  
std               7.443719          3.515277      2.576879           30.588004  
min             -35.678115          1.000000      1.000000           98.200000  
25%              -5.012312          4.000000      2.750000          149.550000  
50%               0.100000          7.000000      5.000000          169.800000  
75%               1.850000         10.000000      7.000000          188.300000  
max               5.000000         12.000000      9.000000          232.500000

# Machine Learning

## Regresi

In [ ]:
# Menentukan variabel X dan y
X = df.drop(columns=["Weight Change (lbs)"])
y = df["Weight Change (lbs)"]

# Mengubah variabel kategorikal menjadi numerik dengan LabelEncoder
encoder = LabelEncoder()
X.loc[:, 'Physical Activity Level'] = encoder.fit_transform(X['Physical Activity Level'])
X.loc[:, 'Gender'] = encoder.fit_transform(X['Gender'])
X.loc[:, 'Sleep Quality'] = encoder.fit_transform(X['Sleep Quality'])

# Split data menjadi train dan test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Model Linear Regression
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Prediksi pada data test
y_pred = linear_model.predict(X_test)

# Hasil regres1
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error (MSE):", mse)
print("R-squared:", r2)

Mean Squared Error (MSE): 0.0004829581244021959
R-squared: 0.9999633494852612


**Interpretasi:**

Hasil dari model Linear Regression menunjukkan performa yang sangat baik dalam memprediksi data. Dengan **Mean Squared Error (MSE)** sebesar 0.00048, yang menunjukkan bahwa selisih antara prediksi model dan nilai aktual sangat kecil, model ini mampu memberikan prediksi yang sangat akurat. Selain itu, nilai **R-squared** yang sangat tinggi, yaitu 0.99996, mengindikasikan bahwa model mampu menjelaskan hampir seluruh variasi dalam data, dengan hanya sedikit kesalahan yang terjadi. Hal ini menunjukkan bahwa model Linear Regression bekerja dengan sangat baik dalam memprediksi variabel target berdasarkan variabel x. Sehingga, model dapat dianggap sangat efektif pada data ini.

## Klasifikasi

In [ ]:
# Mengubah 'Weight Change (lbs)' menjadi kategori (penurunan, peningkatan, stabil)
df['Weight Change Category'] = df['Weight Change (lbs)'].apply(
    lambda x: 'Penurunan' if x < 0 else ('Peningkatan' if x > 0 else 'Stabil')
)

# Menyusun variabel x dan y
X = df[['Stress Level', 'Sleep Quality', 'Duration (weeks)', 'Daily Calories Consumed', 'Physical Activity Level', 'BMR (Calories)']]
y = df['Weight Change Category']

# Mengubah variabel kategorikal menjadi numerik dengan LabelEncoder
encoder = LabelEncoder()
X.loc[:, 'Physical Activity Level'] = encoder.fit_transform(X['Physical Activity Level'])
X.loc[:, 'Sleep Quality'] = encoder.fit_transform(X['Sleep Quality'])

# Stratified KFold dengan 5 fold
stratified_kfold = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

### Naive Bayes

In [ ]:
# Membuat dan melatih model Naive Bayes
nb_model = GaussianNB()

# Evaluasi model
nb_scores = cross_val_score(nb_model, X, y, cv=stratified_kfold, scoring='accuracy')
print("Cross-validation scores Naive Bayes:", nb_scores)
print("Mean cross-validation score:", nb_scores.mean())
print("Standard deviation of score:", nb_scores.std())

Cross-validation scores Naive Bayes: [0.9  0.84]
Mean cross-validation score: 0.87
Standard deviation of score: 0.030000000000000027


**Interpretasi :**

Berdasarkan hasil evaluasi model Naive Bayes diatas, hasil yang di dapatkan cukup memuaskan. Model mencapai akurasi sebesar **90% (0.9)** pada fold pertama dan **84% (0.84)** pada fold kedua. Rata-rata akurasi keseluruhan adalah **87% (0.87)** yang menunjukkan kemampuan model dalam memprediksi data dengan baik. Variasi antara fold cukup kecil dengan standar deviasi **0.03**, yang berarti model relatif konsisten dalam menghasilkan hasil yang baik di setiap fold.

### Decision Trees

In [ ]:
# Membuat dan melatih model Decision Tree
dt_model = DecisionTreeClassifier(random_state=42)

# Evaluasi model
dt_scores = cross_val_score(dt_model, X, y, cv=stratified_kfold, scoring='accuracy')
print("Cross-validation scores Decision Tree:", dt_scores)
print("Mean cross-validation score:", dt_scores.mean())
print("Standard deviation of score:", dt_scores.std())

Cross-validation scores Decision Tree: [1. 1.]
Mean cross-validation score: 1.0
Standard deviation of score: 0.0


Interpretasi:

Berdasarkan hasil evaluasi model Decision Tree di atas menunjukkan performa yang sangat baik dengan nilai sebesar **1.0** pada semua fold. Hal ini menunjukkan bahwa model mampu memprediksi data dengan akurasi sempurna pada setiap subset data. Selain itu, standar deviasi sebesar **0.0** menunjukkan bahwa performa model konsisten di semua fold tanpa variasi.

Namun, hasil ini perlu dilakukan pengecekan lebih lanjut karena akurasi yang terlalu tinggi dan tanpa variasi dapat menjadi indikasi **overfitting**, di mana model mungkin terlalu cocok dengan data pelatihan (train) sehingga tidak dapat menggeneralisasi dengan baik pada data baru.

### Support Vector Machines (SVM)

In [ ]:
# Membuat dan melatih model SVM
svm_model = SVC(random_state=42)

# Evaluasi model
svm_scores = cross_val_score(svm_model, X, y, cv=stratified_kfold, scoring='accuracy')
print("Cross-validation scores SVM:", svm_scores)
print("Mean cross-validation score:", svm_scores.mean())
print("Standard deviation of score:", svm_scores.std())

Cross-validation scores SVM: [0.5  0.52]
Mean cross-validation score: 0.51
Standard deviation of score: 0.010000000000000009


**Interpretasi:**

Berdasarkan hasil evaluasi model Support Vector Machines (SVM) di atas, meskipun model menunjukkan konsistensi dengan variasi yang kecil (standar deviasi **0.01**), hasil akurasi yang dicapai cukup rendah, yaitu **50% (0.5)** pada fold pertama dan **52% (0.52)** pada fold kedua. Rata-rata akurasi keseluruhan adalah **51% (0.51)**, yang menunjukkan bahwa model hanya sedikit lebih baik daripada tebakan acak. Meskipun model SVM ini konsisten dalam hal performa antara fold, ia kurang efektif dalam memprediksi kelas-kelas target dengan baik, sehingga model ini dapat dianggap kurang tepat untuk dataset ini.

### Random Forest

In [ ]:
# Membuat dan melatih model Random Forest
rf_model = RandomForestClassifier(random_state=42)

# Evaluasi model
rf_scores = cross_val_score(rf_model, X, y, cv=stratified_kfold, scoring='accuracy')
print("Cross-validation scores Random Forest:", rf_scores)
print("Mean cross-validation score:", rf_scores.mean())
print("Standard deviation of score:", rf_scores.std())

Cross-validation scores Random Forest: [0.98 0.94]
Mean cross-validation score: 0.96
Standard deviation of score: 0.020000000000000018


**Interpretasi:**

Berdasarkan hasil evaluasi model Random Forest, model ini menunjukkan performa yang sangat baik. Pada fold pertama, model mencapai akurasi **98% (0.98)**, dan pada fold kedua **94% (0.94)**, dengan rata-rata akurasi keseluruhan **96% (0.96)**. Angka ini menunjukkan bahwa Random Forest mampu mengklasifikasikan data dengan sangat baik dan memiliki kemampuan generalisasi yang tinggi. Selain itu, standar deviasi **0.02** menunjukkan bahwa variasi antar fold sangat kecil, yang berarti model cukup stabil dan konsisten. Secara keseluruhan, Random Forest merupakan model yang sangat efektif dan dapat diandalkan untuk klasifikasi data ini.

### K-Nearest Neighbors (KNN)

In [ ]:
# Membuat dan melatih model KNN
knn_model = KNeighborsClassifier()

# Evaluasi model
knn_scores = cross_val_score(knn_model, X, y, cv=stratified_kfold, scoring='accuracy')
print("Cross-validation scores KNN:", knn_scores)
print("Mean cross-validation score:", knn_scores.mean())
print("Standard deviation of score:", knn_scores.std())

Cross-validation scores KNN: [0.48 0.46]
Mean cross-validation score: 0.47
Standard deviation of score: 0.009999999999999981


**Interpretasi:**

Berdasarkan hasil evaluasi model K-Nearest Neighbors (KNN), model ini menunjukkan performa yang kurang baik. Pada fold pertama, model hanya mencapai akurasi **48% (0.48)**, dan pada fold kedua **46% (0.46)**, dengan rata-rata akurasi keseluruhan **47% (0.47)**. Akurasi yang rendah ini menunjukkan bahwa model KNN kesulitan dalam mengklasifikasikan data dengan benar. Lalu, standar deviasi **0.01** menunjukkan variasi yang sangat kecil antar fold, yang berarti model tetap konsisten dalam hasilnya meskipun performanya tidak optimal. Secara keseluruhan, model KNN tidak berhasil memberikan hasil yang memadai untuk klasifikasi data ini, sehingga model ini dapat di anggap kurang tepat untuk dataset yang ini.

## Interpretasi

Secara keseluruhan, hasil cross-validation menunjukkan bahwa model **Decision Tree** dan **Random Forest** adalah yang terbaik dalam menangani data ini. Decision Tree memberikan hasil sempurna dengan akurasi **100%** pada setiap fold, meskipun hal ini bisa menunjukkan adanya risiko **overfitting**, yaitu model terlalu cocok dengan data pelatihan sehingga kemungkinan besar tidak dapat digeneralisasi dengan baik pada data yang baru. Sementara itu, Random Forest menghasilkan akurasi yang sangat tinggi **96%** dengan variasi yang relatif kecil, menunjukkan bahwa model ini cukup stabil dan memiliki kemampuan generalisasi yang baik.

Model **Naive Bayes** juga menunjukkan performa yang cukup baik, dengan rata-rata akurasi **87%**. Meskipun tidak sebaik Random Forest, Naive Bayes tetap memberikan hasil yang stabil dan cukup efektif dalam memprediksi kelas. Namun, model **SVM** dan **KNN** menunjukkan hasil yang kurang memuaskan, dengan akurasi hanya sekitar **51%** dan **47%**. Hal ini menunjukkan bahwa kedua model tersebut kurang efektif dalam menangani dataset ini, bahkan mungkin hanya sedikit lebih baik dari tebakan acak. Secara keseluruhan, model terbaik yang bisa dipilih adalah **Random Forest**, yang memberikan akurasi tinggi dengan stabilitas yang baik, sementara **Decision Tree** juga bisa digunakan, namun perlu hati-hati terhadap potensi **overfitting**.